Source/AssetRipper.Processing/AnimationClips/Editor/StreamedCurveKey.cs

```diff
@@ -12,96 +10,91 @@ public StreamedCurveKey(int index, Vector3 coefficient, float value)
+				float deltaX = nextTime - time;
+				float _CoeffX = Coefficient.X * deltaX * deltaX;
+				float _CoeffY = Coefficient.Y * deltaX;
+				nextKey.inSlope = 3 * _CoeffX + 2 * _CoeffY + Coefficient.Z;
+				//nextKey.inSlope = float.Round(nextKey.inSlope, 4); // amount of decimal places may be more dynamic
+				if (nextKey.Coefficient == default) // calculate LSV only if next slope calculation needs it
+				{
+					nextKey.LSV = (_CoeffX + _CoeffY + Coefficient.Z) * deltaX + Value;
+				}
+				return;
 			}
-			if (deltaTime < 0.00009999999747378752)
+			if (Value != LSV)
 			{
-				deltaTime = 0.000099999997f;
+				// check percentage difference
+				float diff = Value > LSV ? Value - LSV : LSV - Value;
+				float AbsVal = Value < 0 ? -Value : Value;
+				float AbsLSV = LSV < 0 ? -LSV : LSV;
+				float div = AbsVal > AbsLSV ? AbsVal : AbsLSV;
+				float ROUNDING_ERROR = 1e-5f;
+				// normally the difference between Value and LSV should be "big"/much greater than a rounding error.
+				// if the check fails and INCORRECTLY skips this IF block, the next statements
+				// will still produce a good approximation for the expected curve
+				if (diff/div > ROUNDING_ERROR)
+				{
+					outSlope = float.NegativeInfinity;
+					nextKey.inSlope = 0f;
+					nextKey.LSV = Value;
+					Value = LSV;
+					return;
+				}
 			}
-			float deltaValue = nextValue - Value;
-			float inverseTime = 1.0f / (deltaTime * deltaTime);
-			float outTangent = OutSlope * deltaTime;
-			float inTangent = deltaValue + deltaValue + deltaValue - outTangent - outTangent - Coefficient.Y / inverseTime;
-			return inTangent / deltaTime;
+			if (Value == nextKey.Value)
+			{
+				outSlope = 0f;
+				nextKey.inSlope = 0f;
+				nextKey.LSV = Value;
+				return;
+			}
+			outSlope = float.PositiveInfinity;
+			nextKey.inSlope = 0f;
+			// don't do nextKey.LSV=Value here, because having 2 consecutive keys
+			// with outSlope +Inf and -Inf is illegal (Editor corrects it)
 		}
 
-		public float OutSlope => Coefficient.Z;
-
+		/// <summary>
+		/// Index for its GenericBinding inside AnimationClip's BindingConstant.
+		/// </summary>
 		public int Index { get; set; }
+		/// <summary>
+		/// Coefficients of the Cubic Bezier Equation between this Key and the next, for the same Curve.
+		/// </summary>
 		public Vector3 Coefficient { get; set; }
+		/// <summary>
+		/// Value of the key, when approached from its right side.
+		/// </summary>
+		/// <remarks>
+		/// outSlope=-Infinity creates a discontinuity on the curve,
+		/// making the CurveKey value differ from its left side.
+		/// </remarks>
 		public float Value { get; set; }
+		/// <summary>
+		/// Value of the key, when approached from its left side
+		/// </summary>
+		/// <remarks>
+		/// outSlope=-Infinity creates a discontinuity on the curve,
+		/// making the CurveKey value differ from its right side.
+		/// </remarks>
+		public float LSV { get; set; }
+		public float inSlope = 0;
+		public float outSlope = 0;
```

Comment:
Except in rare situations, fields should never be public. You probably want `public float Property { get; private set; }`.